### PR TITLE
Search and Sort on Uplifted Refchains

### PR DESCRIPTION
**Pull request by jamesagnew**

### CODE REVIEW

File: hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/registry/SearchParameterCanonicalizer.java [hunk 1]
```diff
@@ -362,19 +363,30 @@ private RuntimeSearchParam canonicalizeSearchParameterR4Plus(IBaseResource theNe
...
 					}
 				}
 			}
 		}
 	}
 
+	@SuppressWarnings("unchecked")
+	private void addUpliftRefchain(RuntimeSearchParam theRuntimeSearchParam, IBaseExtension<? extends IBaseExtension, ?> theExtension) {
+		String code = ExtensionUtil.extractChildPrimitiveExtensionValue(theExtension, HapiExtensions.EXTENSION_SEARCHPARAM_UPLIFT_REFCHAIN_PARAM_CODE);
+		String elementName = ExtensionUtil.extractChildPrimitiveExtensionValue(theExtension, HapiExtensions.EXTENSION_SEARCHPARAM_UPLIFT_REFCHAIN_ELEMENT_NAME);
```
Comment (michaelabuckley) [inline hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/registry/SearchParameterCanonicalizer.java:384]: I never see `elementName` used.  Is it for comments?  Or what?

File: hapi-fhir-jpaserver-searchparam/src/test/java/ca/uhn/fhir/jpa/searchparam/registry/SearchParamRegistryImplTest.java [hunk 2]
```diff
@@ -338,6 +343,38 @@ public void testAddActiveSearchparam() {
 		assertEquals("FOO", value.getValueAsString());
 	}
 
+	@Test
+	public void testUpliftRefchains() {
+		SearchParameter sp = new SearchParameter();
+		Extension upliftRefChain = sp.addExtension().setUrl(HapiExtensions.EXTENSION_SEARCHPARAM_UPLIFT_REFCHAIN);
+		upliftRefChain.addExtension(HapiExtensions.EXTENSION_SEARCHPARAM_UPLIFT_REFCHAIN_PARAM_CODE, new CodeType("name1"));
+		upliftRefChain.addExtension(HapiExtensions.EXTENSION_SEARCHPARAM_UPLIFT_REFCHAIN_ELEMENT_NAME, new StringType("element1"));
```
Comment (michaelabuckley) [inline hapi-fhir-jpaserver-searchparam/src/test/java/ca/uhn/fhir/jpa/searchparam/registry/SearchParamRegistryImplTest.java:351]: I still don't get what EXTENSION_SEARCHPARAM_UPLIFT_REFCHAIN_ELEMENT_NAME is for.  This test ignores it.

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/MultitenantBatchOperationR4Test.java [hunk 3]
```diff
@@ -250,7 +250,7 @@ public void testReindexByUrl() {
 		List<String> alleleObservationIds = reindexTestHelper.getAlleleObservationIds(myClient);
 		// Only the one in the first tenant should be indexed
 		myTenantClientInterceptor.setTenantId(TENANT_A);
-		MatcherAssert.assertThat(reindexTestHelper.getAlleleObservationIds(myClient), hasSize(1));
+		await().until(() -> reindexTestHelper.getAlleleObservationIds(myClient), hasSize(1));
```
Comment (michaelabuckley) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/MultitenantBatchOperationR4Test.java:253]: issue:  This doesn't look right.  You are polling here, but `alleleObservationIds` is still the original value.  Why poll at all then?  And isn't the reindex done?  So what is this polling doing.  Needs a comment at least.

File: hapi-fhir-jpaserver-test-r4b/src/test/java/ca/uhn/fhir/jpa/dao/r4b/FhirResourceDaoR4BSearchNoFtTest.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class FhirResourceDaoR4BSearchNoFtTest extends BaseJpaR4BTest {
+
+	/**
+	 * Index for
+	 *   [base]/Bundle?composition.patient.identifier=foo
+	 */
+	@ParameterizedTest
+	@CsvSource({"urn:uuid:5c34dc2c-9b5d-4ec1-b30b-3e2d4371508b", "Patient/ABC"})
```
Comment (michaelabuckley) [inline hapi-fhir-jpaserver-test-r4b/src/test/java/ca/uhn/fhir/jpa/dao/r4b/FhirResourceDaoR4BSearchNoFtTest.java:32]: trivial: single values a safer coming from @Values instead of @CsvSource to avoid parsing risk.

**Review by michaelabuckley — APPROVED** [review]
Neat stuff.  The javadoc was very helpful.

It would be nice to standarize on language between chained vs uplifted.

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/context/RuntimeSearchParam.java [hunk 5]
```diff
@@ -358,4 +396,9 @@ public String getReference() {
 		}
 	}
 
+
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/context/RuntimeSearchParam.java:399]: Whoops - deleted

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/util/ExtensionUtil.java [hunk 6]
```diff
@@ -291,4 +295,29 @@ public static void setExtension(FhirContext theFhirContext, IBase theBase, Strin
...
+
+	/**
+	 * Given an extension, looks for the first child extension with the given URL of {@literal theChildExtensionUrl}
+	 * and a primitive datatype value, and returns the String version of that value
+	 *
+	 * @param theExtension The parent extension. Must not be null.
+	 * @param theChildExtensionUrl The child extension URL. Must not be null or blank.
+	 * @since 6.6.0
+	 */
+	public static <D, T extends IBaseExtension<T, D>> String extractChildPrimitiveExtensionValue(@Nonnull IBaseExtension<T, D> theExtension, @Nonnull String theChildExtensionUrl) {
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/util/ExtensionUtil.java:307]: Good call, added.

This method gets well exercised indirectly by other tests but there is nothing checking it's edge conditions. added.

File: hapi-fhir-jpaserver-searchparam/src/test/java/ca/uhn/fhir/jpa/searchparam/registry/SearchParamRegistryImplTest.java [hunk 7]
```diff
@@ -338,6 +343,38 @@ public void testAddActiveSearchparam() {
 		assertEquals("FOO", value.getValueAsString());
 	}
 
+	@Test
+	public void testUpliftRefchains() {
+		SearchParameter sp = new SearchParameter();
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-jpaserver-searchparam/src/test/java/ca/uhn/fhir/jpa/searchparam/registry/SearchParamRegistryImplTest.java:348]: heh... i can go either way there. Agree the java syntax is not great

File: hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/registry/SearchParameterCanonicalizer.java (hunk 1 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/registry/SearchParameterCanonicalizer.java:384]: It's used in the Mongo implementation only. I'm gonna move us over to using this codebase for that part in a separate ticket.

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/MultitenantBatchOperationR4Test.java (hunk 3 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/MultitenantBatchOperationR4Test.java:264]: You're right - this was me chasing down a false lead in an intermittent. I'm rolling it back, that was not the issue.

File: hapi-fhir-jpaserver-test-r4b/src/test/java/ca/uhn/fhir/jpa/dao/r4b/FhirResourceDaoR4BSearchNoFtTest.java (hunk 4 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-test-r4b/src/test/java/ca/uhn/fhir/jpa/dao/r4b/FhirResourceDaoR4BSearchNoFtTest.java:32]: I was not aware of @ValueSource. Nice, added.

**Comment by jkiddo** [comment]
@jamesagnew will Location.near also be a valid parameter to this `_sort`?
